API8 fixes

Faithcaio opened the pull request:
a bunch of fixes for the API8 implementation

Faithcaio committed:
fix SqlServiceImpl

Faithcaio committed:
fix incompatible types compile errors

Faithcaio committed:
fix SpongeTriggerBuilder missing from()

Faithcaio committed:
Block DyeColors

Faithcaio committed:
fix most EntityUtil compile errors
Teleport code still needs a thorough check

Faithcaio committed:
fix Tracker PhaseStates

Faithcaio committed:
fix PacketPhaseUtil main hand slot restoration

Faithcaio committed:
fix SpongeEntityDamageSourceBuilder

Faithcaio committed:
notifier is data

gabizou commented on src/main/java/org/spongepowered/common/event/damage/SpongeEntityDamageSourceBuilder.java:
```diff
@@ -49,8 +49,8 @@ public SpongeEntityDamageSourceBuilder entity(final Entity entity) {
         checkState(this.reference.get() != null);
-        final net.minecraft.util.EntityDamageSource damageSource =
-            new net.minecraft.util.EntityDamageSource(this.damageType.getId(), (net.minecraft.entity.Entity) this.reference.get());
+        final net.minecraft.util.EntityDamageSource damageSource = // TODO damageType String is used for the translation key!
```
this is negotiated based on the mixin  and damage source to type provider.

Faithcaio committed:
fix SmeltingRecipe mixin

Faithcaio committed:
remove SpongeUser raw data

Faithcaio committed:
a bunch of easy mixin fixes

gabizou commented on src/accessors/java/org/spongepowered/common/accessor/entity/player/ServerPlayerEntityAccessor.java:
```diff
@@ -22,33 +22,15 @@
-        }
-    }
+@Mixin(ServerPlayerEntity.class)
+public interface ServerPlayerEntityAccessor {
```
Do we not even have one of these from before?

Faithcaio replied:
I dont see another accessor for ServerPlayerEntity

gabizou commented on src/main/java/org/spongepowered/common/bridge/block/DyeColorBlockBridge.java:
```diff
@@ -26,7 +26,10 @@
+public interface DyeColorBlockBridge {
 
     void bridge$setDyeColor(DyeColor dyeColor);
+    @Nullable DyeColor bridge$getDyeColor();
```
I'd rather
```suggestion
    Optional<DyeColor> bridge$getDyeColor();
```

gabizou commented on src/main/java/org/spongepowered/common/data/provider/DataProviderRegistry.java:
```diff
@@ -100,10 +101,10 @@ private static boolean filterHolderType(DataProvider<?,?> provider, Class<?> hol
-                .map(key -> this.getProvider(key, holderType))
-                .filter(provider -> !(provider instanceof EmptyDataProvider))
-                .collect(Collectors.toMap(DataProvider::getKey, Function.identity())));
+        Stream<DataProvider> stream = this.dataProviders.keySet().stream()
```
```suggestion
        final Stream<DataProvider> stream = this.dataProviders.keySet().stream()
```

gabizou commented on src/main/java/org/spongepowered/common/data/provider/DataProviderRegistry.java:
```diff
@@ -100,10 +101,10 @@ private static boolean filterHolderType(DataProvider<?,?> provider, Class<?> hol
+        Stream<DataProvider> stream = this.dataProviders.keySet().stream()
+                .map(key -> this.getProvider((Key) key, holderType))
+                .filter(provider -> !(provider instanceof EmptyDataProvider));
+        return new DataProviderLookup(stream.collect(Collectors.toMap(p -> (Key<?>) p.getKey(), p -> (DataProvider<?, ?>) p)));
```
This is really janky to read. Either collect on the stream above or declare a new variable to pass in.

Faithcaio replied:
Collection on the stream caused incompatible types compile errors

gabizou commented on src/main/java/org/spongepowered/common/data/provider/DataProviderRegistry.java:
```diff
@@ -143,10 +144,9 @@ public DataProviderLookup getProviderLookup(Class<?> dataHolderType) {
-                .map(key -> buildDelegateProvider(key, (List) this.dataProviders.get(key).stream().filter(predicate)))
-                .collect(Collectors.toMap(DataProvider::getKey, Function.identity()));
-        return new DataProviderLookup(dataProviders);
+        Stream<DataProvider> stream = this.dataProviders.keySet().stream()
```
final

gabizou commented on src/main/java/org/spongepowered/common/data/provider/DataProviderRegistry.java:
```diff
@@ -143,10 +144,9 @@ public DataProviderLookup getProviderLookup(Class<?> dataHolderType) {
-        return new DataProviderLookup(dataProviders);
+        Stream<DataProvider> stream = this.dataProviders.keySet().stream()
+                .map(key -> buildDelegateProvider((Key) key, (List) this.dataProviders.get(key).stream().filter(predicate)));
+        return new DataProviderLookup(stream.collect(Collectors.toMap(p -> (Key<?>) p.getKey(), p -> (DataProvider<?, ?>) p)));
```
Same as before, janky to read..

gabizou commented on src/main/java/org/spongepowered/common/data/provider/item/stack/ItemStackDyeColorProvider.java:
```diff
@@ -40,17 +41,19 @@ public ItemStackDyeColorProvider() {
     @Override
     protected Optional<DyeColor> getFrom(ItemStack dataHolder) {
-        return ((DyeColorItemBrige)dataHolder).bridge$getDyeColor();
+        Item item = dataHolder.getItem();
+        if (item instanceof BlockItem) {
+            Block block = ((BlockItem) item).getBlock();
+            net.minecraft.item.DyeColor color = ((DyeColorBlockBridge) block).bridge$getDyeColor();
+            return Optional.ofNullable((DyeColor) (Object) color);
```
```suggestion
        final Item item = dataHolder.getItem();
        if (item instanceof BlockItem) {
            final Block block = ((BlockItem) item).getBlock();
            return ((DyeColorBlockBridge) block)..bridge$getDyeColor();
```
This way you have the optional already provided by the bridge implementation

gabizou commented on src/main/java/org/spongepowered/common/data/util/StateHelper.java:
```diff
@@ -35,7 +35,7 @@
         for (final Map.Entry<IProperty<?>, Comparable<?>> entry : from.getValues().entrySet()) {
             if (original.has(entry.getKey())) {
-                original = (S) original.with(entry.getKey(), (Comparable) entry.getValue());
+                original = (S) original.with((IProperty) entry.getKey(), (Comparable) entry.getValue());
```
Gonna need a class level suppress warnings for raw usage

gabizou commented on src/main/java/org/spongepowered/common/entity/EntityUtil.java:
```diff
@@ -406,72 +407,69 @@ public static ServerPlayerEntity transferPlayerToWorld(final ServerPlayerEntity
             // Check if the world we're going to matches our provider type. if so, we need to send a fake respawn packet to clear chunks
-            final int fromClientDimId = WorldManager.getClientDimensionId(player, fromWorld);
-            final int toClientDimId = WorldManager.getClientDimensionId(player, toWorld);
+            DimensionType fromType = fromWorld.getDimension().getType();
+            DimensionType toType = toWorld.getDimension().getType();
```
```suggestion
            final DimensionType fromType = fromWorld.getDimension().getType();
            final DimensionType toType = toWorld.getDimension().getType();
```

gabizou commented on src/main/java/org/spongepowered/common/entity/EntityUtil.java:
```diff
@@ -406,72 +407,69 @@ public static ServerPlayerEntity transferPlayerToWorld(final ServerPlayerEntity
-            if (fromClientDimId == toClientDimId) {
-                final int fakeDimId;
-                switch (fromClientDimId) {
+            if (fromType == toType) {
```
@Zidane is this right with dimension handling? I'd imagine the type could be the same but the dimension id assigned by the WorldManager could be different?

Faithcaio replied:
This is to send a fake respawn packet to vanilla clients in case the dimension type is not changing.
This entire teleport things needs extensive testing anyways once we can actually start a server.
Maybe its not needed anymore?

Zidane replied:
No you still have to fake sent a dimension change else you must send an unload packet for every chunk.

That is not including the other fun things the client sets based on the dimension.

i509VCB replied:
I think this is still mentioned in the protocol wiki

gabizou commented on src/main/java/org/spongepowered/common/entity/EntityUtil.java:
```diff
@@ -406,72 +407,69 @@ public static ServerPlayerEntity transferPlayerToWorld(final ServerPlayerEntity
-            if (fromWorld.dimension.getType() == DimensionType.NETHER && toWorld.dimension.getType() == DimensionType.OVERWORLD
-                && player.getEnteredNetherPosition() != null) {
-                CriteriaTriggers.NETHER_TRAVEL.trigger(player, player.getEnteredNetherPosition());
+            Vec3d enteredNetherPosition = ((ServerPlayerEntityAccessor) player).accessor$getEnteredNetherPosition();
```
```suggestion
            final Vec3d enteredNetherPosition = ((ServerPlayerEntityAccessor) player).accessor$getEnteredNetherPosition();
```

gabizou commented on src/main/java/org/spongepowered/common/entity/EntityUtil.java:
```diff
@@ -520,30 +516,22 @@ private static InvokingTeleporterContext createInvokingTeleporterPhase(final Ent
         // Check if we're to use a different teleporter for this world
         if (teleporter.getClass().getName().equals("net.minecraft.world.Teleporter")) {
-            worldName = portalAgents.get("minecraft:default_" + toWorld.dimension.getType().getName().toLowerCase(Locale.ENGLISH));
+            CatalogKey key = ((DimensionTypeBridge) toWorld.dimension.getType()).bridge$getKey();
```
```suggestion
            final CatalogKey key = ((DimensionTypeBridge) toWorld.dimension.getType()).bridge$getKey();
```

gabizou commented on src/main/java/org/spongepowered/common/entity/EntityUtil.java:
```diff
@@ -661,18 +649,19 @@ public static boolean isEntityDead(final net.minecraft.entity.Entity entity) {
     }
 
     public static MoveEntityEvent.Teleport handleDisplaceEntityTeleportEvent(final Entity entityIn, final double posX, final double posY, final double posZ, final float yaw, final float pitch) {
+        World world = ((org.spongepowered.api.entity.Entity) entityIn).getWorld();
```
```suggestion
        final World world = ((org.spongepowered.api.entity.Entity) entityIn).getWorld();
```

gabizou commented on src/main/java/org/spongepowered/common/event/tracking/phase/plugin/ListenerPhaseContext.java:
```diff
@@ -43,7 +44,7 @@
     private CapturePlayer capturePlayer;
 
-    ListenerPhaseContext(final IPhaseState<L> state) {
+    ListenerPhaseContext(final IPhaseState<L> state, PhaseTracker tracker) {
```
```suggestion
    ListenerPhaseContext(final IPhaseState<L> state, final PhaseTracker tracker) {
```

gabizou reviewed (changes requested):
Turn on local variable can be final inspection and fix them. There's also some areas where it just looks really long or nested method calls that should be local variables (the compiler can take care of inlining).

Faithcaio committed:
pr comments

Faithcaio committed:
fix furnace toContainer

Faithcaio committed:
fix jukebox mixin

Faithcaio committed:
fix sign TE mixin

Faithcaio committed:
fix SoundEvent Mixin

Faithcaio committed:
move invalid Text mixins

Faithcaio committed:
fix dyecolor again

Faithcaio committed:
fix FallingBlock mixin

Faithcaio committed:
fix FarmlandBlock mixin

Faithcaio committed:
fix FireBlock mixin targets

Faithcaio committed:
fix TNTBlock mixin

Faithcaio committed:
fix SpongeUser mixin
other DataCompoundHolder impls are gone?

Faithcaio committed:
fix ProjectileDispenseBehavior mixin

Faithcaio committed:
fix ArmorStandEntity overwrite

Faithcaio committed:
not invalid anymore

Faithcaio committed:
partial fix CreeperEntity mixin

Faithcaio committed:
fix FishingBobberEntity mixin

Faithcaio committed:
fix projectile mixins
trying to fix explosion mixins (smoking parameter does not exist)

Faithcaio committed:
was this a performance fix?

Faithcaio committed:
fix SScoreboardObjectivePacket mixin

Faithcaio committed:
mostly fix Scoreboard mixins

Faithcaio committed:
fix BeaconTileEntity mixin

Faithcaio committed:
Update/Fix tileentity ticking optimization based on PaperMC patch:

https://github.com/PaperMC/Paper/blob/master/Spigot-Server-Patches/0022-Optimize-TileEntity-Ticking.patch

and move it to the optimization module

Faithcaio committed:
fix HopperTileEntity mixin

Faithcaio committed:
removed it

Faithcaio committed:
fix some Text mixins

Faithcaio committed:
fix id -> key usage

Faithcaio committed:
fix entity collision mixins

Faithcaio committed:
fix WorkbenchContainerMixin

Faithcaio committed:
fix ItemFrameEntityMixin_Optimization_Map

Faithcaio committed:
fix FilledMapItem_Optimization_Map

Faithcaio committed:
fix collision optimization mixinw

Faithcaio committed:
fix real time mixins

Faithcaio committed:
fix ServerPlayerEntityMixin_Tracker

Faithcaio committed:
fix tracker mixins
@gabizou unsure about the ChunkManagerMixin_Tracker
where the target is a lambda
